sdl-core/src/main/scala/io/smartdatalake/workflow/action/generic/transformer/DeduplicateTransformer.scala

```diff
@@ -0,0 +1,71 @@
+/*
+ * Smart Data Lake - Build your data lake the smart way.
+ *
+ * Copyright © 2019-2023 ELCA Informatique SA (<https://www.elca.ch>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package io.smartdatalake.workflow.action.generic.transformer
+
+import com.typesafe.config.Config
+import io.smartdatalake.config.SdlConfigObject.{ActionId, DataObjectId}
+import io.smartdatalake.config.{FromConfigFactory, InstanceRegistry}
+import io.smartdatalake.definitions.Environment
+import io.smartdatalake.util.hdfs.PartitionValues
+import io.smartdatalake.workflow.action.Action
+import io.smartdatalake.workflow.dataframe.GenericDataFrame
+import io.smartdatalake.workflow.dataobject.TableDataObject
+import io.smartdatalake.workflow.{ActionPipelineContext, DataFrameSubFeed}
+
+/**
+ * Returns dataframe with only unique columns based on a primary key
+ * @param name              Name of the transformer
+ * @param description       Optional description of the transformer
+ * @param rankingExpression Ranking expression to determine duplicates
+ * @param primaryKeyColumns Optional list of primary key columns
+ */
+case class DeduplicateTransformer(override val name: String = "DeduplicateTransformer", override val description: Option[String] = None, rankingExpression: String, primaryKeyColumns: Option[Seq[String]] = Option.empty[Seq[String]]) extends GenericDfTransformer {
+
+  override def transform(actionId: ActionId, partitionValues: Seq[PartitionValues], df: GenericDataFrame, dataObjectId: DataObjectId, previousTransformerName: Option[String], executionModeResultOptions: Map[String, String])(implicit context: ActionPipelineContext): GenericDataFrame = {
+
+
+    val functions = DataFrameSubFeed.getFunctions(df.subFeedType)
+    import functions._
+
+    val primaryKeys: Option[Seq[String]] = primaryKeyColumns.orElse(
+      Some(Environment.instanceRegistry.get[Action](objectId = actionId)).getOrElse(Option.empty[Action]) match {
+        case action: Action => action.outputs.head match {
+          case dataObject: TableDataObject => dataObject.table.primaryKey
+          case _ => Option.empty[Seq[String]]
+        }
```

Comment:
Some(...).getOrElse(Option.empty) looks strange to me, as Some is always defined...
The following should work as well and is much simpler, right?
Additionally we can assume Action is of subtype DataFrameActionImpl and use the mainOutputId, which makes the transformer a little bit more generic.
```suggestion
      val action = Environment.instanceRegistry.get[DataFrameActionImpl](objectId = actionId)
      val output = action.outputs.find(_.id == action.mainOutputId).get
      output match {
        case dataObject: TableDataObject => dataObject.table.primaryKey
        case _ => Option.empty[Seq[String]]
      }
```

Reply:
Changed to Some to Option b379c4411638354604fc209b609480c5ea478afb

Reply:
It doesnt matter, Option(...:Action).getOrElse(Option.empty[Action]) does not make sense. Its even not a proper type afterwards, because you resolve the first option but return again None[Action] from getOrElse.

I rewrote it as follows:
```
    val primaryKey: Seq[String] = primaryKeyColumns.orElse {
      val action = Environment.instanceRegistry.get[DataFrameActionImpl](actionId)
      action.mainOutput match {
        case dataObject: TableDataObject => dataObject.table.primaryKey
        case _ => None
      }
    }.getOrElse(throw ConfigurationException("There are no primary key columns defined ether by parameter nor by detection with actionId."))
```